File: RNColorLibrary/ios/AtomicRnColorLibrary.m
#import <React/RCTBridgeModule.h>
#import <React/RCTViewManager.h>

@interface RCT_EXTERN_MODULE(AtomicRnColorLibrary, NSObject)

RCT_EXTERN_METHOD(generateColor:(nonnull NSString)typeString withRed:(nonnull NSNumber)red withGreen:(nonnull NSNumber)green withBlue:(nonnull NSNumber)blue withResolver:(RCTPromiseResolveBlock)resolve withRejecter:(RCTPromiseRejectBlock)reject)

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

@end

File: RNColorLibrary/ios/AtomicRnColorLibrary.swift
import RGColorSDK
import Foundation

@objc(AtomicRnColorLibrary)
class AtomicRnColorLibrary: NSObject {
    
    @objc(generateColor:withRed:withGreen:withBlue:withResolver:withRejecter:)
    func generateColor(_ typeString: NSString,
                       red: NSNumber,
                       green: NSNumber,
                       blue: NSNumber,
                       resolve: RCTPromiseResolveBlock,
                       reject: RCTPromiseRejectBlock) {
        
        // TODO: add error handling
        let colorType: ColorType
        if typeString == "hex" {
            colorType = .hex
        } else {
            colorType = .rgb
        }
        guard let redDouble = red as? Double,
              let blueDouble = blue as? Double,
              let greenDouble = green as? Double
        else {
            resolve("#FF00FF")
            return
        }
        
        let red = validateColorValue(redDouble)
        let blue = validateColorValue(blueDouble)
        let green = validateColorValue(greenDouble)
        let colorString = ColorCalculator.generateColor(colorType: colorType, redIn: red, greenIn: green, blueIn: blue)
        
        resolve(colorString)
    }
    
    private func validateColorValue(_ value: Double) -> Double? {
        if value < 0 || value > 255 {
            return nil
        } else {
            return value
        }
    }
}


